book/visual-effects.md

```diff
@@ -648,338 +631,305 @@ of thumb is: if you don't need a non-default blend mode, then you can use
+`Pixel` class.
 
 ``` {.python file=examples}
-# Note: this is sample code to explain the concept, it is not part
-# of the actual browser.
-def composite(source_color, backdrop_color, compositing_mode):
-    if compositing_mode == "source-over":
-        (source_r, source_g, source_b, source_a) = \
-            tuple(source_color)
-        (backdrop_r, backdrop_g, backdrop_b, backdrop_a) = \
-            tuple(backdrop_color)
-        return skia.Color4f(
-            backdrop_r * (1-source_a) * backdrop_a + \
-                source_r * source_a,
-            backdrop_g * (1-source_a) * backdrop_a + \
-                source_g * source_a,
-            backdrop_b * (1-source_a) * backdrop_a + \
-                source_b * source_a,
-            1 - (1 - source_a) * (1 - backdrop_a))
+class Pixel:
+    def __init__(self, r, g, b, a):
+        self.r = r
+        self.g = g
+        self.b = b
+        self.a = a
 ```
 
-[^other-compositing]: We'll shortly encounter other compositing modes.
+When we apply a `Paint` with an `Alphaf` parameter, the first thing
+Skia does is add the requested opacity to each pixel:
 
-This algorithm is basically "average the source and backdrop pixel colors,
-weighted by their their alpha".^[Examples: if the alpha of the source pixel
-was 1, the result is just the source pixel color, and if it was 0 the
-result is the backdrop pixel color.]
+``` {.python file=examples}
+class Pixel:
+    def alphaf(self, opacity):
+        self.a = self.a * opacity
+```
 
-Putting it all together, if we were to implement the `restore` command
-ourselves from one canvas to another, we could write the following
-(pretend we have
-a `getPixel` method that returns a `skia.Color4f` and a `setPixel` one that
-sets a pixel color):[^real-life-reading-pixels]
+I want to emphasize that this code is not a part of our browser---I'm
+simply using Python code to illustrate what Skia is doing internally.
 
-``` {.python.example}
-# Note: this is sample code to explain the concept, it is not part
-# of the actual browser.
-def restore(source_surface, backdrop_surface, width, height, opacity):
-    for x in range(0, width):
-        for y in range(0, height):
-            backdrop_surface.setPixel(
-                x, y,
-                composite(
-                    apply_opacity(source_surface.getPixel(x, y)),
-                    backdrop_surface.getPixel(x, y)))
-```
+That `Alphaf` operation applies to pixel in one surface. But with
+`SaveLayer` we will end up with two surfaces, with all of their pixels
+aligned, and therefore we will need to combined, or *blend* pairs of
+corresponding pixels.
+
+Here the terminology can get confusing: we imagine that the pixels "on
+top" are blending into the pixels "below", so we call the top surface
+the *source surface*, with source pixels, and the bottom surface the
+*destination surface*, with destination pixels. When we combine them,
```

Comment:
I called it backdrop to conform to the spec. If destination is more clear, we can use that. But backdrop
does have an intuitive meaning maybe?

Reply:
Yeah I could go either way here. "Destination" is nice because then "destination-in" makes a lot of sense. But "backdrop" is nice because it reminds you which one is the backdrop. We could also introduce both terms, but that itself can be confusing.

Reply:
ok let's go with destination because of dst-in.